src/replica/DirectorIndexJob.cc

```diff
@@ -477,54 +414,117 @@ void IndexJob::_processRequestData(util::Lock const& lock, IndexRequest::Ptr con
+            // succesfully finishing the ingest into the destination table.
+            _completeChunks++;
+
+        } catch (exception const& ex) {
+            string const error = context_ + "failed to load data into the 'director' index table, ex: " +
+                                 string(ex.what());
+            LOGS(_log, LOG_LVL_ERROR, error);
+            if (state() == State::FINISHED) return;
+            util::Lock lock(_mtx, context_);
+            if (state() == State::FINISHED) return;
+            _resultData.error[request->worker()][request->chunk()] = error;
+            finish(lock, ExtendedState::FAILED);
+            return;
         }
-        default:
-            throw range_error(typeName() + "::" + string(__func__) +
-                              "  unsupported destination: " + toString(_destination));
     }
 }
 
-list<IndexRequest::Ptr> IndexJob::_launchRequests(util::Lock const& lock, string const& worker,
-                                                  size_t maxRequests) {
-    list<IndexRequest::Ptr> requests;
+DirectorIndexRequest::Ptr DirectorIndexJob::_nextRequest() {
+    string const context_ = context() + string(__func__) + " ";
+    LOGS(_log, LOG_LVL_DEBUG, context_);
+
+    // This method has to use std::condition_variable::wait_for()
+    // instead of std::condition_variable::wait() in order to allow rechecking
+    // changes in a status of the job, not just notifications on new entries
+    // in the queue _completedRequests.
+    //
+    // The re-check interval of 1s should be reasonable here since this type
+    // of jobs is designed for processing large amounts of data, which may
+    // take many minutes or even many hours.
+    chrono::milliseconds const jobStatusCheckIvalMsec{1s};
+
+    // Pull the next request from the queue or wait.
+    DirectorIndexRequest::Ptr request;
+    while (request == nullptr) {
+        if (state() == State::FINISHED) return nullptr;
+        unique_lock<mutex> lock(_mtx);
+        if (state() == State::FINISHED) return nullptr;
+
+        if (!_completedRequests.empty()) {
+            request = _completedRequests.front();
+            _completedRequests.pop_front();
+            _numLoadingRequests++;
+        } else {
+            _cv.wait_for(lock, jobStatusCheckIvalMsec, [&] {
+                if (state() == State::FINISHED) return true;
+                if (_completeChunks == _totalChunks) {
+                    // We got here because the main condition for completing the job
+                    // has been met. It means the job needs to be finished right away.
+                    // Unfortunately, the desired state transition can't be done under
+                    // the currently held lock of thr type std::unique_lock. The API for
+                    // the operation require acquering a different type of locks
+                    // represented by the class util::Lock.
```

Comment:
The timed version of `wait_for` is needed here to prevent the method from hanging forever in the case when no more work exists for the MySQL loading threads. And the loop is needed for handling the so-called "spurious" unlocks of `wait_for` when the  index builder job is still alive and more work is expected for threads that called the method:
```
DirectorIndexRequest::Ptr DirectorIndexJob::_nextRequest() {
```
As far as the locking is concerned, there is no need to pass the lock around in this context. The unique lock is acquired in the local context in coordination with calling `wait_for` on the condition variable as needed to pull work from the job's queue and pass the work orders to the calling threads.